server/auth/types/authentication_type.ts

```diff
@@ -38,8 +46,176 @@ export type IAuthHandlerConstructor = new (
+      }
+    } else {
+      // no auth header in request, try cookie
+      try {
+        cookie = await this.sessionStorageFactory.asScoped(request).get();
+      } catch (error) {
+        this.logger.error(`Error parsing cookie: ${error.message}`);
+        cookie = undefined;
+      }
+
+      if (!cookie || !this.isValidCookie(cookie)) {
+        // clear cookie
+        this.sessionStorageFactory.asScoped(request).clear();
+        // send to auth workflow
+        return this.redirectToAuth(request, response, toolkit);
+      }
+
+      // extend cookie expiration time
+      cookie!.expiryTime = Date.now() + this.config.cookie.ttl;
+      this.sessionStorageFactory.asScoped(request).set(cookie!);
+
+      // cookie is valid
+      // build auth header
+      const authHeadersFromCookie = this.buildAuthHeaderFromCookie(cookie!);
+      Object.assign(authHeaders, authHeadersFromCookie);
+      const additonalAuthHeader = this.getAdditionalAuthHeader(request);
+      Object.assign(authHeaders, additonalAuthHeader);
+    }
+
+    // resolve tenant if necessary
+    if (this.config.multitenancy?.enabled && isMultitenantPath(request)) {
+      try {
+        const tenant = await this.resolveTenant(request, cookie!, authHeaders, authInfo);
+        // return 401 if no tenant available
+        if (!tenant) {
+          return response.badRequest({
+            body:
+              'No availabe tenant for current user, please reachout to your system administrator',
+          });
+        }
+        // set tenant in header
+        Object.assign(authHeaders, { securitytenant: tenant });
+
+        // set tenant to cookie
+        if (tenant !== cookie!.tenant) {
+          cookie!.tenant = tenant;
+          this.sessionStorageFactory.asScoped(request).set(cookie!);
+        }
+      } catch (error) {
+        this.logger.error(`Failed to resolve user tenant: ${error}`);
+        if (error instanceof UnauthenticatedError) {
+          return this.redirectToAuth(request, response, toolkit);
+        }
+        throw error;
+      }
+    }
+
+    return toolkit.authenticated({
+      requestHeaders: authHeaders,
+    });
+  };
+
+  isRestApiCall(request: KibanaRequest) {
+    return (
+      request.headers[AuthenticationType.REST_API_CALL_HEADER] &&
+      (request.headers[AuthenticationType.REST_API_CALL_HEADER] as string).toLocaleLowerCase() ===
+        'true'
+    );
+  }
+
+  authNotRequired(request: KibanaRequest) {
+    const pathname = request.url.pathname;
+    // allow access to assets
+    if (pathname && pathname.startsWith('/bundels/')) {
+      return true;
+    }
+    // allow requests to ignored routes
+    if (pathname && AuthenticationType.ROUTES_TO_IGNORE.includes(pathname)) {
+      return true;
+    }
+    // allow requests to routes that doesn't require authentication
+    if (pathname && this.config.auth.unauthenticated_routes.indexOf(pathname) > -1) {
+      // TODO: use kibana server user
+      return true;
+    }
```

Comment:
Minor: instead of checking pathname on every if, probably you can add a

if(!pathname){
 return false;
}

Reply:
agreed